Port `max` kernel to structured kernels.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #61449

Tracking issue: #55070

Differential Revision: [D29741714](https://our.internmc.facebook.com/intern/diff/D29741714)

ysiraichi committed:
Port `max` kernel to structured kernels.

Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `max` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `max` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi commented:
These 3 CI failures don't seem related to this PR.

bdhirsh commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -397,58 +420,34 @@ std::tuple<Tensor &,Tensor &> mode_out(const Tensor& self, int64_t dim, bool kee
-  return result;
+std::tuple<Tensor, Tensor> max(const Tensor& self, int64_t dim, bool keepdim) {
+  // CUDA and CPU dispatch keys are handled by the structured kernel implementation.
+  TORCH_INTERNAL_ASSERT(self.is_quantized());
```
I think it's probably fine to take out the internal assert, since we don't assert dispatcher correctness in our other native kernels.

Also maybe name this kernel to something like `qmax`, since it's only used for quantized now?

bdhirsh commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -397,58 +420,34 @@ std::tuple<Tensor &,Tensor &> mode_out(const Tensor& self, int64_t dim, bool kee
-      Tensor max = at::empty({0}, self.options());
-      return at::native::max_out(self, dim, keepdim, max, max_indices);
+      auto& values_mut = const_cast<Tensor&>(values);
+      auto& indices_mut = const_cast<Tensor&>(indices);
```
Is there a reason to use `const_cast` here instead of updating the types of `max_stub` (and the corresponding `max_kernel_impl` code in `TensorCompareKernel.cpp`) to use `const Tensor&` everywhere?

(If updating that signature requires changing a ton of call-sites then the const_cast is probably fine)

ysiraichi committed:
Update on "Port `max` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `max` kernel to structured kernels."


Tracking issue: #55070

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29741714).

ezyang commented on aten/src/ATen/native/ReduceOpsUtils.h:
```diff
@@ -298,6 +295,18 @@ static void zero_numel_tensor_resize(Tensor& result, Tensor& result_indices,
+// This function should be called when you are reducing a zero-numel tensor and want to
+// resize the output and return it. This function exists for resizing zero-numel
+// tensors when the size of the reduction dimension is non-zero.
+static void zero_numel_tensor_resize(Tensor& result, Tensor& result_indices,
```
you probably should take result as `const Tensor&` here

ysiraichi replied:
This function already existed (apparently, `at::mode` is using it with non-const result). What I did was to extract `get_zero_numel_tensor_size` out of it. Should I change it nonetheless?

ezyang replied:
You don't have to if you don't want to in this PR, but `const Tensor&` argument will work anywhere `Tensor&` used to work, and will save a const cast from a site where you only have a `const Tensor&`. It is not a big deal but just letting you know what the terminal state should look like.

ysiraichi replied:
Ah, sorry. You are right. Since `at::mode` is the only function using it, I think I might as well get rid of this function.

ezyang commented:
@bdhirsh this LGTM, I'll let you land it

ysiraichi committed:
Update on "Port `max` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29741714](https://our.internmc.facebook.com/intern/diff/D29741714)

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `max` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29741714](https://our.internmc.facebook.com/intern/diff/D29741714)

[ghstack-poisoned]

ysiraichi commented:
@bdhirsh I have removed `zero_numel_tensor_resize`, and replaced it by `get_zero_numel_tensor_size` (only user was in `at::mode`, besides `max` and `min`). The CI failure there is not related to this PR.

bdhirsh committed:
Update on "Port `max` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29741714](https://our.internmc.facebook.com/intern/diff/D29741714)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29741714).

ysiraichi committed:
Update on "Port `max` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29741714](https://our.internmc.facebook.com/intern/diff/D29741714)

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `max` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29741714](https://our.internmc.facebook.com/intern/diff/D29741714)

[ghstack-poisoned]

bdhirsh committed:
Update on "Port `max` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29741714](https://our.internmc.facebook.com/intern/diff/D29741714)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29741714).

bdhirsh committed:
Update on "Port `max` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29741714](https://our.internmc.facebook.com/intern/diff/D29741714)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29741714).

bdhirsh committed:
Update on "Port `max` kernel to structured kernels."


Tracking issue: #55070

Differential Revision: [D29741714](https://our.internmc.facebook.com/intern/diff/D29741714)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29741714).
